[DRAFT] EIP-3978: Introduce gas refunds on reverts (#3978)
ethereum/EIPs: EIPS/eip-3978.md

---
eip: 3978
title: Gas refunds on reverts
description: Do not erase gas refunds on transaction subcall reverts, due users pay a lot of gas for storage non-modification.
author: Anton Bukov (@k06a), Mikhail Melnik (@ZumZoom)
discussions-to: https://ethereum-magicians.org/t/eip-3978-gas-refunds-on-reverts/7071/
status: Draft
type: Standards Track
category: Core
created: 2021-09-16
---

## Abstract

> Since [EIP-3298](https://eips.ethereum.org/EIPS/eip-3298) gas refunds works for storage restores only inside the same transaction. For example [ERC-20](./eip-20.md) `approve` + `transferFrom` flow between 2 smart contracts according to [EIP-2200](https://eips.ethereum.org/EIPS/eip-2200) and [EIP-2929](https://eips.ethereum.org/EIPS/eip-2929) will cost nearly to `21600` gas with gas refund counter `20000`. But in case of reverting this subcall (containing both `approve` and `transferForm`) gas refund will be erased, while smart contract storage will remain unmodified. I think it should keep storage access costs, but still refund modification costs.

Reviewer: ```suggestion
 Since [EIP-3298](./eip-3298.md) gas refunds works for storage restores only inside the same transaction. For example [ERC-20](./eip-20.md) `approve` + `transferFrom` flow between 2 smart contracts according to [EIP-2200](./eip-2200.md) and [EIP-2929](./eip-2929.md) will cost nearly to `21600` gas with gas refund counter `20000`. But in case of reverting this subcall (containing both `approve` and `transferForm`) gas refund will be erased, while smart contract storage will remain unmodified. I think it should keep storage access costs, but still refund modification costs.
```
Note that the reference to EIP-3298 will cause this EIP to depend on 3298.  Consider removing that reference if you don't want this EIP to be held up by the progress of that EIP.

## Motivation

Сurrent full cancelling of gas refunds on internal reverts is too unfair. Users pay for non-modification same cost as for modification.

## Specification

> Let's count all SSTORE gas costs within every subcall, excluding access costs. And on reverting any subcall let's not erase refund counter, but:

Reviewer: This needs to be much more clearly specified, and it needs to be phrased as an assertion not a suggestion (don't use terms like "let's" as it comes off as a suggestion rather than an assertion.  EIPs should be technical specifications that assert how things MUST behave.

Author: Thanks, this makes sense.

```
tx.gas_refund_counter = tx.gas_refund_counter - call.gas_refund_counter + MIN(call.gas_refund_counter, call.all_sstores_gas_cost)
```

## Rationale

TBD

## Backwards Compatibility

> Should not break anything since gas refunds are only applied at the end of the transaction.

Reviewer: As mentioned above, this section should be written using more assertive language.  "Should not break anything" does not instill confidence in the reader.  A common statement for this section is:
```suggestion
No known backward incompatibilities.
```

---
eip: 3978
title: Gas refunds on reverts
author: Anton Bukov (@k06a), Mikhail Melnik (@ZumZoom)

> discussions-to: <to be assigned>

Reviewer 2: ```suggestion
discussions-to: https://ethereum-magicians.org/t/eip-3978-gas-refunds-on-reverts/7071/2
```

status: Draft
type: Standards Track
category: Core
created: 2021-09-16
---

> ## Simple Summary

Reviewer 2: This section has been replaced with the `description` field in the preamble. Please see the `eip-template.md` for more info.

> Do not zeroify gas refunds of transaction subcall reverts, due users pay a lot of gas for storage non-modification.

Reviewer 2: I'm a little confused by this sentence?

    
## Abstract

> >A short (~200 word) description of the technical issue being addressed. This should be a very terse and human-readable version of the specification section. Someone should be able to read only the abstract to get the gist of what this specification does.

Reviewer 2: ```suggestion
```

> Since [EIP-3298](https://eips.ethereum.org/EIPS/eip-3298) gas refunds works for storage restores only inside the same transaction. For example ERC20 `approve` + `transferFrom` flow between 2 smart contracts according to [EIP-2200](https://eips.ethereum.org/EIPS/eip-2200) and [EIP-2929](https://eips.ethereum.org/EIPS/eip-2929) will cost nearly to `21600` gas with gas refund counter `20000`. But in case of reverting this subcall (containing both `approve` and `transferForm`) gas refund will be erased, while smart contract storage will remain unmodified. I think it should keep storage access costs, but still refund modification costs.

Reviewer 2: ```suggestion
Since [EIP-3298](https://eips.ethereum.org/EIPS/eip-3298) gas refunds works for storage restores only inside the same transaction. For example [ERC-20](./eip-20.md) `approve` + `transferFrom` flow between 2 smart contracts according to [EIP-2200](https://eips.ethereum.org/EIPS/eip-2200) and [EIP-2929](https://eips.ethereum.org/EIPS/eip-2929) will cost nearly to `21600` gas with gas refund counter `20000`. But in case of reverting this subcall (containing both `approve` and `transferForm`) gas refund will be erased, while smart contract storage will remain unmodified. I think it should keep storage access costs, but still refund modification costs.
```

## Motivation

> >The motivation section should describe the "why" of this EIP. What problem does it solve? Why should someone want to implement this standard? What benefit does it provide to the Ethereum ecosystem? What use cases does this EIP address?

Reviewer 2: Please remove all the template comments

> I find current full cancel of gas refunds because of internal reved too unfair. Users pay for non-modification same cost as for modification.

Reviewer 2: Please avoid using a first person narrative voice in EIPs.
